전남대 BE_이현기 2주차 과제 (1단계)

geist9110 opened the pull request:
- 맨 처음에 의도한 구현 기능 목록과 실제로 구현하는 도중에 잘 맞지 않는 부분이 생겨서 중간에 수정을 했습니다.
- 원래 의도는 KakaoStringValidation, SpecialCharacterValidation에서 각각 별도의 커스텀 예외를 처리할 수 있을 줄 알고 진행을 했었습니다. 하지만 직접 커스텀 예외를 발생시키려면 각각의 validator 내부에서 false가 나오는 부분에 throw를 직접 사용해서 커스텀 예외를 발생시키는 방법 말고는 찾지 못했습니다.
- 레포지토리를 옮기는 도중에 문제가 생긴거 같습니다. 0단계 이후의 commit은 [cc424b7](https://github.com/kakao-tech-campus-2nd-step2/spring-gift-wishlist/pull/161/commits/cc424b78d4e273215083b2419e6a4638de799330) 부터입니다.

wotjd243 committed:
Initial commit

wotjd243 committed:
feat: set up the project

LeeHyungi-geist9110 committed:
docs: Add list of implemented features

LeeHyungi-geist9110 committed:
docs: Add detail explanation of implementation feature i1

LeeHyungi-geist9110 committed:
feat(Product): Add ProductClass

LeeHyungi-geist9110 committed:
feat(Product): Add ProductRepository class

ProductRepositroy class has singleton pattern, so create this repository with getInstance method
now if you create this class, then initial data also create
idGenerator make auto increment key

LeeHyungi-geist9110 committed:
feat(Product): Add ProductController class

API PATH : /api/products
GET : parameter Long id
POST : body Product product
PATCH : parameter Long id body Product product
DELETE : parmeter Long id

LeeHyungi-geist9110 committed:
feat: Add ProductResponseDTO

LeeHyungi-geist9110 committed:
refactor(Product): Remove ProductResponseDTO class and Add id variable in Product class

product class need id variable.
if product class does not have id variable, the code becomes too complicated

LeeHyungi-geist9110 committed:
refactor(Product): Remove initial data and Change all method return changed product

LeeHyungi-geist9110 committed:
refactor(Product): Remove ResponseEntity and get all values with RequestParam

LeeHyungi-geist9110 committed:
docs(Admin): Change list of implemented features for step2

LeeHyungi-geist9110 committed:
docs(Admin): Add admin page example image

LeeHyungi-geist9110 committed:
feat(Admin): Add thymeleaf dependency and properties

LeeHyungi-geist9110 committed:
feat(Admin): Add adminPage.html and adminPage.css

Admin controller is not connected not yet

LeeHyungi-geist9110 committed:
feat(Admin): Connect with server and view products each

LeeHyungi-geist9110 committed:
feat(Admin): Connect with thymleaf in adminPage Controller

LeeHyungi-geist9110 committed:
feat(Admin): Add modal in Add a new product button

LeeHyungi-geist9110 committed:
style(Admin): Design modal and split adminPage.css to adminPage.css and modal.css

LeeHyungi-geist9110 committed:
feat(Admin): Add addProduct function

LeeHyungi-geist9110 committed:
feat(Admin): Add product checkbox work

if you click first checkbox, then all of checkbox will be checked
also, if you check all of checkbox, then first checkbox will be checked

LeeHyungi-geist9110 committed:
style(Admin): Add comments on each component

LeeHyungi-geist9110 committed:
feat(Admin): Add delete checked row function and resultUrlAndReload function

when select some checkbox and click delete button, then those are deleted in database
resultUrlAndReload function is used for clean url
when add product or delete some products then url change like  /api/products?name= ...
so, this function can remove parameters and reload admin page

LeeHyungi-geist9110 committed:
feat(Admin): Add delete product function for each product and refactor delete checked product

LeeHyungi-geist9110 committed:
feat(Admin): Add edit button use product modal

Add new Product button changed h1 tag and submit button when click add new product button

LeeHyungi-geist9110 committed:
feat(Admin): Add pagination feature

LeeHyungi-geist9110 committed:
feat(Admin): Add pagination attribute for thymeleaf

LeeHyungi-geist9110 committed:
refactor(Admin): Change pagination to requestparam from pathvariable

LeeHyungi-geist9110 committed:
feat(Admin): Add page swap with index

LeeHyungi-geist9110 committed:
refactor(Admin): Refactor AdminPageController

Split const values(PAGE_SIZE, MAX_PAGE_INDEX) to AdminPageConfigure class
Split getPageListRange method

LeeHyungi-geist9110 committed:
refactor(Admin): Change adminPage.js to module

all function changed export function ... style
remove not used variables

LeeHyungi-geist9110 committed:
refacor(Admin): Split adminPage.js and productAPI.js

adminPage.js controle button click
productAPI.js controle API related with product class

LeeHyungi-geist9110 committed:
refactor(Admin): Rename ambiguous variables

LeeHyungi-geist9110 committed:
docs(Database): Write step3 list of implementation feature

LeeHyungi-geist9110 committed:
feat(Database): Add properties for h2database

LeeHyungi-geist9110 committed:
feat(Database): Add createProductTable

LeeHyungi-geist9110 committed:
feat(Database): Add addProduct method with jdbc

LeeHyungi-geist9110 committed:
feat(Database): Add getAllProduct method with jdbc

LeeHyungi-geist9110 committed:
feat(Database): Add updateProduct method with jdbc

LeeHyungi-geist9110 committed:
feat(Database): Add deleteProduct method with jdbc

LeeHyungi-geist9110 committed:
docs(E2E): Wirte list of complementation features for step4

LeeHyungi-geist9110 committed:
feat(E2E): Add selenium and webdriver dependency

LeeHyungi-geist9110 committed:
test(E2E): Add product add e2e test case

LeeHyungi-geist9110 committed:
refactor: Rename getAllProduct to getAllProducts

LeeHyungi-geist9110 committed:
refactor: Change id value input from pathvariable

LeeHyungi-geist9110 committed:
refactor: Change create table method to schema.sql

LeeHyungi-geist9110 committed:
refactor: Remove delete product method's return value

LeeHyungi-geist9110 committed:
refactor: Remove addProduct method's return value

LeeHyungi-geist9110 committed:
style: Relocation private method

LeeHyungi-geist9110 committed:
style: Relocation getAllProducts parameters

LeeHyungi-geist9110 committed:
style: Remove not used paramter in addProduct method

LeeHyungi-geist9110 committed:
Merge remote-tracking branch 'gift-product-repository/main'

LeeHyungi-geist9110 committed:
docs(validation): Write list of implementation features for wishlist step1

LeeHyungi-geist9110 committed:
chore(validation): Add validation dependency

LeeHyungi-geist9110 committed:
refactor: Add ProductDTO class and change input from request param to request body

remove addProduct method's return value
remove updateProduct method's return value

add ProductDTO class that include name, price, imageUrl
addProduct and updateProduct method get request body(ProductDTO)

addProduct method in ProductRepository modified more simply

productAPI is also changed addProduct and editProduct by server's API
rename apiUrl to productAPIUrl

LeeHyungi-geist9110 committed:
style: Seperate hierachical structure

Relocation related Product class to Product folder
Relocation related Admin to Admin folder

LeeHyungi-geist9110 committed:
feat(validation): Add product name input validation

product name's length is must be between 1 and 15

LeeHyungi-geist9110 committed:
feat(validation): Add product name input special chracter validation

all of special chracter detected in SpecialCharacterValidator and find not allowed characters in while loop

LeeHyungi-geist9110 committed:
feat(validation): Add product name input 'kakao' validation

if product name has kakao then return false

LeeHyungi-geist9110 committed:
style: Relocation pattern and modify typo in KaKao

LeeHyungi-geist9110 committed:
feat(validation): Add MethodArgumentNotValidException error handler

LeeHyungi-geist9110 committed:
feat: Add ProductName invalid input view

clien can show what is problem in product name when client add product invalid input

LeeHyungi-geist9110 committed:
feat(validation): Add productDTO's length validation message

LeeHyungi-geist9110 committed:
feat(validation): Add "카카오" string validation

LeeHyungi-geist9110 committed:
doc(validation): Rewrite list of implementation features for wishlist step1

LeeHyungi-geist9110 committed:
style(validation): Change kakaoStringValidation's default message

geist9110 committed:
Merge branch 'geist9110' into main

soominsohn commented on README.md:
```diff
@@ -9,10 +9,11 @@
+      - 클라이언트는 제품을 입력하는 칸 밑에 "( ), [ ], +, -, &, /, _" 이외의 특수문자는 허용되지 않습니다. 라는 코멘트를 볼 수 있다.
+   3. 제품 이름에 "카카오"라는 단어가 포함된 값이 입력된 값은 허용하지 않는다.
+      - `400 BAD REQUEST`를 발생시킨다.
+      - 클라이언트는 제품을 입력하는 칸 밑에 "담당 MD와 협의한 경우에만 사용할 수 있습니다." 라는 코멘트를 볼 수 있다.
```
HTTP Response code를 바꾸셨군요! 💯 
어떤 이유로 400 BAD REQUEST로 바꾸셨나요?

geist9110 replied:
처음엔 각 예외를 제가 세밀하게 컨트롤할 수 있으면 좋을거 같다는 생각에 각 검증 절차마다 별도의 Status 코드와 예외를 지정해줬습니다. 하지만 조금 더 생각해보니 같은 검증 절차일 뿐인 코드에 각기 다른 에러를 발생시키면 복잡해질 뿐만 아니라 적절한 Status 코드가 아니라는 생각이 들어 수정하게 되었습니다.

soominsohn commented on src/main/java/gift/Validation/SpecialCharacterValidator.java:
```diff
@@ -0,0 +1,36 @@
+            if (!ALLOWED_SPECIAL_CHARACTER.contains(matcher.group())) {
+                return false;
+            }
+        }
```
정규표현식을 수정하면 `KakaoStringValidator`처럼 로직이 간단해질 것 같습니다. 정규표현식을 사용하면 표현이 복잡해지므로 주석을 다는 것이 좋습니다. 어차피 주석을 달 예정이라면, 허용하는 문자와 허용하지 않는 문자를 하나의 정규표현식으로 작성하는 것이 isValid 함수의 구현을 간단하게 만들어 가독성을 높일 수 있습니다. :)

soominsohn commented on src/main/java/gift/Product/Product.java:
```diff
@@ -0,0 +1,4 @@
+package gift.Product;
+
+public record Product(Long id, String name, Integer price, String imageUrl) {
```
price를 int가 아닌 Integer를 사용하셨는데요, 
java의 primitive 타입과 wrapper 타입에 대해 공부해보고 어떤 자료형이 더 적절할지 고민해보아요 😄 
참고: https://velog.io/@kimdy0915/%EA%B8%B0%EB%B3%B8%ED%98%95primitive-vs.-%EB%9E%98%ED%8D%BC-%ED%81%B4%EB%9E%98%EC%8A%A4wrapper-class

geist9110 replied:
Wrapper 타입을 사용하는 것이 조금 더 바람직하다 생각합니다. Wrapper 타입을 통해 Collection 프레임워크를 사용하면 비지니스 로직에서 stream을 사용해 로직의 성능을 조금 더 끌어올릴 수 있다고 생각합니다. 또한 primitive 타입으로 선언할 경우, 비지니스 로직이나 다른 부분에서 한번 더 Wrapper 타입으로 변환해줘야 하는 불필요한 부분이 생길것으로 판단됩니다.

soominsohn replied:
Wrapper클래스의 장점에 대해 잘 공부하신 것 같아요. 👍
그러나 보통 primitive를 많이 사용합니다. primitive type을 사용하면 객체를 생성하지 않아 메모리 효율측면에서도 좋고, 불필요한 오토박싱/언박싱을 피할 수 있습니다. 또한 Wrapper 클래스를 사용하게 되면 null값을 허용하게 되는데 예외처리와 로직을 복잡하게 만듭니다. 
말씀해주신 Wrapper 타입을 이용해 Collection 프레임워크를 이용하는 경우에만 Wrapper 타입을 사용하시는게 좋고, 그 외의 경우는 primitive type을 사용하는게 적절합니다.

soominsohn commented on src/main/java/gift/Product/ProductController.java:
```diff
@@ -0,0 +1,44 @@
+
+    @PostMapping("/api/products")
+    public void addProduct(
+            @Valid @RequestBody(required = true) ProductDTO productDto) {
```
맥북을 사용하신다면 `Command + B`,
윈도우를 사용하신다면 `Ctrl + B`로 @RequestBody 어노테이션의 구현으로 들어갈 수 있습니다. 
들어가보시면 required의 default 값이 true로 되어있는 것을 확인하실 수 있어요 :)
default값이 명시되어 있다면 굳이 적어주지 않으셔도 됩니다!
```
@Target({ElementType.PARAMETER})
@Retention(RetentionPolicy.RUNTIME)
@Documented
public @interface RequestBody {
    boolean required() default true;
}
```

soominsohn commented on src/main/java/gift/Product/ProductDTO.java:
```diff
@@ -0,0 +1,16 @@
+import gift.Validation.KakaoStringValidation;
+import gift.Validation.SpecialCharacterValidation;
+
+public record ProductDTO(
```
이번 미션에서 DTO를 추가하셨군요. 
Product와 ProductDTO의 차이는 무엇이며 각각 어떤 역할을 할까요?

geist9110 replied:
Product는 DB에 들어가는 도메인 값 자체를 의미합니다. 이는 DB의 각 열들과 1-1로 맵핑되어 값을 행별로 원본 그대로 가져와주는 역할을 합니다.
ProductDTO는 사용자와의 상호작용을 하기 위한 객체를 의미합니다. 이는 Controller에서 RequestBody를 통해 사용자에게 값을 입력받는 역할을 수행합니다.
둘의 차이는 ID값의 존재 유무입니다. 사용자가 상품을 추가할 때 ID값을 직접 입력하지 않기에 ProductDTO는 ID값이 필요 없습니다. 반면에 Product는 DB의 각 열들과 1-1 맵핑이 되어야 하기에 ID값이 반드시 필요합니다.

soominsohn replied:
Product와 ProcutDTO의 차이에 대해 잘 고민해 주셨네요. 그러나 Product와 ProductDTO의 차이는 단순히 ID 값의 존재 유무로만 구분하는 것 외에 각각의 역할과 책임이 있습니다. 

**도메인 객체**

- Product객체를 도메인 객체라고도 하는데요, 다음과 같은 역할을 합니다.
- 도메인 객체는 데이터베이스와의 매핑뿐만 아니라, 비즈니스 로직을 포함하는 핵심 객체입니다. 예를 들어 product 객체의 필드에 대한 핵심 비즈니스 로직 검증이 이루어집니다.

**DTO**

- DTO는 단순히 계층간 데이터를 전달하기 위한 객체로, 비즈니스 로직이 포함되지 않습니다. DTO에서의 검증은 주로 Bean Validation을 사용해 사용자의 입력이 제대로 들어오는지 정도로만 검증합니다.

https://velog.io/@linger0310/DDD

또, 현기님이 ProductDTO는 사용자와의 상호작용을 하기 위한 객체를 의미라고 말씀해주셨는데 현재 Repository에서도 ProductDTO를 사용하고 있어요. 사용자로부터 전달받은 DTO를 repository레이어에서 까지 사용하는게 적절할까요? DTO가 여러 계층을 넘나드는건 적절하지 않아 보입니다. 수정해보아요 🙂

soominsohn commented on src/main/java/gift/Product/ProductDTO.java:
```diff
@@ -0,0 +1,16 @@
+import gift.Validation.SpecialCharacterValidation;
+
+public record ProductDTO(
+    @Length(min = 1, max = 15, message = "product name's length must be between 1 and 15")
+    @SpecialCharacterValidation
+    @KakaoStringValidation
```
DTO의 name필드에 Bean Validation을 사용하셨군요 👍 
1. 이것은 무엇이고 어떻게 동작하는 걸까요?
2. DTO외에 또 어디서 어떻게 유효값 검증을 해볼 수 있을까요? `Hint: 도메인`
3. 또 CustomValidation을 만들어 주셨는데, 어떤 장단점이 있을까요?

geist9110 replied:
1. 어노테이션들은 바로 밑에 선언되어 있는 첫번째 변수에 대해 검증요청이 들어오면 검증을 진행해줍니다. 이는 각각의 `ConstraintValidator`인터페이스 내부의 isValid 함수를 통과하면 true, 실패하면 false를 반환하고 false가 발생하면 `MethodArgumentNotValidException`을 발생시킵니다.
2. 3가지 위치에서 유효값 검증이 가능합니다. 첫번째로 제가 작성한 것과 같이 DTO에서 검증을 하는 Controller에서 검증을 할 수 있습니다. 두번째로 Service 레이어에서 `@Validated` 어노테이션을 사용해서 비즈니스 로직과 검증 로직을 동시에 실행시킬 수 있습니다. 마지막으로 Repository 레이어에서 DTO를 도메인으로 변경하는 과정에서 도메인의 검증 조건을 사용해서 검증이 가능합니다.
3. 원하는 검증 조건을 직접 만들 수 있고, 추후에 똑같은 검증 조건이 필요할 때 어노테이션으로 손쉽게 재사용이 가능하다는 장점이 있습니다. 단점으로는 직접 만들다보니, 해당 로직을 개발한 사람이 아니면 검증 조건을 정확하게 파악하기 힘들다는 단점이 있습니다.

soominsohn replied:
1,3번에 대해 잘 학습해주셨어요.

2번에 대해 추가적으로 

Controller Layer에서는 DTO의 Bean Validaton을 활용해 사용자의 입력이 제대로 들어오는지 정도(예-null값이 들어지는 않는지, 글자 수가 적절한지 등)을 검증하고요,

Service Layer에서는 DTO를 도메인객체로 변환하고, 도메인 객체(Product)의 생성자/메서드 혹은 서비스 레이어의 메서드에서 비즈니스 로직의 오류를 검증합니다. (예 - 유효하지 않은 특수문자를 사용하지는 않았는지 등)

Service Layer에서 Repository Layer를 호출할 때는 도메인 객체를 다시 Repository Layer용 DTO로 변환해 전달합니다. 하지만 지금 이정도 크기의 프로그램에서는 Product객체를 그대로 전달해줘도 상관 없어 보입니다.

현재 현기님이 작성해주신 코드에서 DTO에 검증 어노테이션을 잘 활용해 주셨고요 장단점도 파악하고 계신 것 같습니다. 

추가적으로 지금 Service Layer는 없지만 추후 생성하게 된다면 Service Layer에서 도메인 객체를 활용한 검증로직 작성도 학습해보고 도전해보면 좋을 것 같습니다.

soominsohn commented on src/main/java/gift/ExceptionHandler/ValidationExcpetionHandler.java:
```diff
@@ -0,0 +1,17 @@
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+
+@ControllerAdvice
```
예외 처리에 @ControllerAdvice를 사용하셨군요!
예외 처리에 사용되는 세가지 어노테이션의 차이점은 무엇일까요?
`@ExceptionHandler`, `@ControllerAdvice`, `@RestControllerAdvice`

geist9110 replied:
1. `@ExceptionHandler`는 어떤 에러에 대해서 컨트롤 할 것인지를 나타냅니다.
2. `@ControllerAdvice`와 `@RestControllerAdvice`는 둘 다 공통적으로 어떤 컨트롤러에 대해서 에러를 컨트롤 할 것인지를 나타내고, 둘의 차이는 Controller와 RestController의 차이처럼 값을 Rest하게 반환하냐 그냥 반환하냐의 차이가 있습니다.
현재 작성한 코드도 Rest하게 보내고 있는데 ControllerAdvice를 사용하고 있으므로, 바로 수정하겠습니다.

soominsohn commented on src/main/java/gift/Product/ProductController.java:
```diff
@@ -0,0 +1,44 @@
+package gift.Product;
```
패키지명은 대문자, 소문자 모두 가능하나 소문자로 많이 씁니다! 👍

soominsohn reviewed (changes requested):
안녕하세요 현기님! 리뷰어 앤지입니다. 
이번 미션도 잘 해주셨네요 💯. 남은 미션도 파이팅입니다!!

작성해주신 코드에 대한 질문과 리뷰 외에 아래 두가지 참고 해주세요. 
* 패키지 분리하시면서 분리하신 패키지 안과 밖에 중복되는 클래스들이 존재해요. 중복된 클래스는 제거해주세요.
* 전체 코드에 대해 정렬을 해주세요. 참고: intelliJ의 Reformat code 
* 또 다음 미션부터는 main 브랜치가 아니라 step2 브랜치에서 PR 부탁드려요.


___
질문에 대한 답변

ConstraintValidator를 이용해 CustomValitaon 잘 만들어 주셨습니다. 👍 
그러나 ConstraintValidator를 구현한 KakaoStringValidation 및 SpecialCharacterValidation에서 커스텀 예외를 발생시키려고 하신 이유가 있을까요?

`ConstraintValidator` 인터페이스를 살펴보면 `boolean isValid(T var1, ConstraintValidatorContext var2);` 메서드가 있습니다. 현기님은 이 메서드를 구현하셨죠.

이 인터페이스를 구현하는 순간부터 예외 처리를 검증 로직이 아닌 시스템에 위임하게 됩니다. 즉, 이 메서드에서 유효성 검사를 수행하고 true 또는 false를 반환하도록 구현하기만 하면 그에 따른 예외 처리는 프레임워크가 알아서 처리하라고 위임하는 것입니다. 실제로 시스템에서 `MethodArgumentNotValidException`이 발생하는 것을 보면 현기님이 코드상에서 직접 이 예외를 던지지 않았음에도 불구하고 시스템이 예외를 처리하고 있는 것을 확인할 수 있습니다.

만약 커스텀 예외를 반환하고 싶다면 `ConstraintValidator`를 구현한 Custom Validator 내에서가 아닌 다른 곳에서 예외 처리를 진행하는 것이 적절해 보입니다. 이 부분은 현기님이 고민해보고 적용해 보시면 좋을 것 같습니다.

LeeHyungi-geist9110 committed:
style(validation): Deleted duplicated files and reformat all file

LeeHyungi-geist9110 committed:
style(validation): Rename typo class and rename query selector

LeeHyungi-geist9110 committed:
refactor(validation): Change ControllerAdvice to ResControllerAdvice

LeeHyungi-geist9110 committed:
refactor(validation): Remove custom validation and add Pattern validation in ProductDTO

validate english, korean, number with special character is more proper than only detect special characters

LeeHyungi-geist9110 committed:
style(validation): Rename module's name to small letter

LeeHyungi-geist9110 committed:
refactor(validation): Separate common API path use RequestMapping

soominsohn reviewed (approved):
현기님! 수정 반영해주신 내용 잘 보았구요, step2를 위해 이만 머지하겠습니다.
코멘트 몇개 남겨놨으니 확인해주시면 감사하겠습니다.
다음 단계도 파이팅이에요 ! 😄
